torch_xla_py/xla_dist.py

```diff
@@ -78,38 +113,158 @@ def validate(self):
+    """Gets the instance group that the current VM belongs to."""
+    resp = self._compute_service.instances().get(
+        project=self._project,
+        zone=self._zone,
+        instance=self._vm_master,
+        fields='metadata').execute()
+
+    if 'metadata' in resp and 'items' in resp['metadata']:
+      for item in resp['metadata']['items']:
+        if (item['key'] == 'created-by' and
+            'instanceGroupManagers' in item['value']):
+          return item['value'].split('/')[-1]
+
+    raise RuntimeError(('A vm list must be passed to ClusterResolver '
+                        'if not using an instance group'))
+
+  def _get_member_instance_names(self, instance_group):
+    """Gets all the instance names that belong to the given instance group."""
+    resp = self._compute_service.instanceGroups().listInstances(
+        project=self._project, zone=self._zone,
+        instanceGroup=instance_group).execute()
+
+    instances = []
+    for item in resp.get('items', []):
+      if 'instance' not in item or 'status' not in item:
+        continue
+      instance_path = item['instance']
+      instances.append(instance_path.split('/')[-1])
+
+    return instances
+
+  def _get_client_workers(self):
+    """Gets client workers.
+
+    The instance group that the current VM belongs to is picked up from
+    the GCE instance metadata set of the VM. If a list of VMs was used for
+    initializing cluster resolver, we use that instead.
+
+    Returns:
+      A list of ClientWorker.
+
+    Raises:
+      RuntimeError: If the red VM cluster is not healthy.
+    """
+    if self._vms is None:
+      # Using an instance group
+      instance_group = self._get_instance_group()
+      self._vms = self._get_member_instance_names(instance_group)
+      if len(self._vms) == 0:
+        raise RuntimeError('Client worker vms is empty in instance group.')
+
+    workers = []
+    batch = self._compute_service.new_batch_http_request()
+
+    def add_worker(request_id, resp, exception):
+      """Callback for each request in BatchHttpRequest."""
+      if exception is not None:
+        raise exception
+      hostname = resp['selfLink'].split('/')[-1]
+      if resp['status'] != 'RUNNING':
+        raise RuntimeError(
+            ('Instance {hostname} is not running yet. '
+             'Re-run when all VMs are running.').format(hostname=hostname))
+      worker = ClientWorker(
+          internal_ip=resp['networkInterfaces'][0]['networkIP'],
+          machine_type=resp['machineType'].split('/')[-1],
+          zone=resp['zone'].split('/')[-1],
+          hostname=hostname)
+      workers.append(worker)
+
+    for vm in self._vms:
+      req = self._compute_service.instances().get(
+          project=self._project,
+          zone=self._zone,
+          instance=vm,
+          fields=('machineType,metadata,selfLink,'
+                  'networkInterfaces/networkIP,status,zone'))
+      batch.add(req, add_worker)
+    batch.execute()
+
```

Comment:
Let's try to keep code more compact, vertical space wise.

Reply:
Did it add an empty line?

Reply:
Empty line after project, zone, instance? Yes.